tick/stateful/functions.go

```diff
@@ -234,6 +256,232 @@ func (m *mathIF) Call(args ...interface{}) (v interface{}, err error) {
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as first arg to %s, must be string", args[0], m.name)
+		return
+	}
+	a1, ok := args[1].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as second arg to %s, must be string", args[1], m.name)
+		return
+	}
+	v = m.f(a0, a1)
+	return
+}
+
+func (m *string2String) Reset() {}
+
+type string1StringFunc func(string) string
+type string1String struct {
+	name string
+	f    string1StringFunc
+}
+
+func newString1String(name string, f string1StringFunc) *string1String {
+	return &string1String{
+		name: name,
+		f:    f,
+	}
+}
+
+func (m *string1String) Call(args ...interface{}) (v interface{}, err error) {
+	if len(args) != 1 {
+		return 0, errors.New(m.name + " expects exactly one argument")
+	}
+	a0, ok := args[0].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as first arg to %s, must be string", args[0], m.name)
+		return
+	}
+	v = m.f(a0)
+	return
+}
+
+func (m *string1String) Reset() {}
+
+type strReplace struct {
+}
+
+func (m *strReplace) Call(args ...interface{}) (v interface{}, err error) {
+	if len(args) != 4 {
+		return 0, errors.New("strReplace expects exactly four arguments")
+	}
+	str, ok := args[0].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as first arg to strReplace, must be string", args[0])
+		return
+	}
+	old, ok := args[1].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as second arg to strReplace, must be string", args[1])
+		return
+	}
+	new, ok := args[2].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as third arg to strReplace, must be string", args[2])
+		return
+	}
+	n, ok := args[3].(int64)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as fourth arg to strReplace, must be int", args[3])
+		return
+	}
+	v = strings.Replace(str, old, new, int(n))
+	return
+}
+
+func (m *strReplace) Reset() {}
+
+type strSubstring struct {
+}
+
+func (m *strSubstring) Call(args ...interface{}) (v interface{}, err error) {
+	if len(args) != 3 {
+		return 0, errors.New("strSubstring expects exactly three arguments")
+	}
+	str, ok := args[0].(string)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as first arg to strSubstring, must be string", args[0])
+		return
+	}
+	start, ok := args[1].(int64)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as second arg to strSubstring, must be int", args[1])
+		return
+	}
+	stop, ok := args[2].(int64)
+	if !ok {
+		err = fmt.Errorf("cannot pass %T as third arg to strSubstring, must be int", args[2])
+		return
+	}
+	if start < 0 {
+		return nil, fmt.Errorf("found negative index for strSubstring: %d", start)
+	}
+	if stop < 0 {
+		return nil, fmt.Errorf("found negative index for strSubstring: %d", stop)
+	}
+	if int(stop) >= len(str) {
+		return nil, fmt.Errorf("stop index too large for string in strSubstring: %d", stop)
+	}
+
+	v = str[start:stop]
+	return
+}
+
+func (m *strSubstring) Reset() {}
+
+type regexReplace struct {
+}
+
+func (m *regexReplace) Call(args ...interface{}) (v interface{}, err error) {
```

Comment:
regexReplace, strSubstring, and strReplace should all have value receivers as well as be used as values in map  above. There is no benefit in storing a pointer to an empty struct. The other types would be fine as value receivers as well, since they only store a string and a function pointer, and are immutable.